pymc3/distributions/distribution.py

```diff
@@ -134,40 +313,127 @@ def __new__(cls, name, *args, **kwargs):
                 "for a standalone distribution."
             )
 
-        rng = kwargs.pop("rng", None)
+        if not isinstance(name, string_types):
+            raise TypeError(f"Name needs to be a string but got: {name}")
 
         if rng is None:
             rng = model.default_rng
 
-        if not isinstance(name, string_types):
-            raise TypeError(f"Name needs to be a string but got: {name}")
-
-        data = kwargs.pop("observed", None)
+        _, dims, _ = _validate_shape_dims_size(dims=dims)
 
-        total_size = kwargs.pop("total_size", None)
+        # Create the RV without specifying testval, because the testval may have a shape
+        # that only matches after replicating with a size implied by dims (see below).
+        rv_out = cls.dist(*args, rng=rng, testval=None, **kwargs)
+        ndim_implied = rv_out.ndim
+        resize_shape = None
 
-        dims = kwargs.pop("dims", None)
+        # `dims` are only available with this API, because `.dist()` can be used
+        # without a modelcontext and dims are not tracked at the Aesara level.
+        if dims is not None:
+            ndim_resize, resize_shape, dims = _resize_from_dims(dims, ndim_implied, model)
+        elif observed is not None:
+            ndim_resize, resize_shape, observed = _resize_from_observed(observed, ndim_implied)
 
-        if "shape" in kwargs:
-            raise DeprecationWarning("The `shape` keyword is deprecated; use `size`.")
+        if resize_shape:
+            # A batch size was specified through `dims`, or implied by `observed`.
+            rv_out = change_rv_size(rv_var=rv_out, new_size=resize_shape, expand=True)
 
-        transform = kwargs.pop("transform", UNSET)
+        if dims is not None:
+            # Register named but previously unknown implied dimensions,
+            # such that they can be used for further downstream.
+            for di, dname in enumerate(dims[ndim_resize:]):
+                if not dname in model.dim_lengths:
+                    model.add_coord(dname, values=None, length=rv_out.shape[ndim_resize + di])
```

Comment:
I understand that it is a bit arbitrary, but it feels  like this would be better placed in `model.register_rv`, no?